src/Storages/StorageMongoDB.cpp

```diff
@@ -51,6 +50,26 @@ StorageMongoDB::StorageMongoDB(
 }
 
 
+void StorageMongoDB::connectIfNotConnected()
+{
+    std::lock_guard lock{connection_mutex};
+    if (!connection)
+        connection = std::make_shared<Poco::MongoDB::Connection>(host, port);
```

Comment:
I think it is better to release mutex right after connection is initialized.

Reply:
It seems there is no point in allowing multiple threads to do the authentication in parallel.

Reply:
ok